### PR TITLE
tracing-policies: support tags to categorize events

### PR DESCRIPTION
**Pull request by tixxdz**
To categorize events we need a "tags" field that is a list of tags.
    
The Sigma Tags Specification is a good starting point:
https://github.com/SigmaHQ/sigma-specification/blob/main/Tags_specification.md

We first define namespaces for events, and we start with the
"observability" namespace.
    
observability: is the ability to collect and export data about the internal system state.

- "observability.filesystem": this event is about filesystem operations.
-  "observability.privilege_escalation": this event is about the rights and permissions
        granted to a process or a user.
- "observability.process": this event is about an instance of a Linux program
        that is being executed.

```
tracing-policies: support tags to categorize events
```

### CODE REVIEW

**Comment by tixxdz** [comment]
@jrfastab the namespace I added is "observability", doing "tracing" is an alternative, strictly speaking they are flexible.

But I prefer to have it "observability.tracing" for things bpf, kprobes, etc instead or even later "integrity.tracing" things that modify the kernel or memory... be more into users/tools that will consume Tetragon events and filter those events by tags, compared to doing the developer way of "tracing.filesystem" or "tracing.x" which will still stay a bit low level and hard to consume from a higher perspective.

**Comment by tixxdz** [comment]
> Thanks that's a good addition! I just have some remarks on the Go code that could be simplified I think to be easier to read.
> 
> I'm not fan of the naming `observability.privilege` though I don't think very talkative while privilege_escalation or whatever is from MITRE or other standards is more explicit.

privilege_escalation seems what other tools are using, it is explicit, with the right filters we can surely use it, let's see what other think

File: pkg/sensors/tracing/generickprobe.go [hunk 1]
```diff
@@ -687,6 +690,11 @@ func addKprobe(funcName string, f *v1alpha1.KProbeSpec, in *addKprobeIn) (id idt
 		logger.GetLogger().WithField("policy-name", in.policyName).Warnf("TracingPolicy 'message' field too long, truncated to %d characters", TpMaxMessageLen)
 	}
 
+	tagsField, err := getPolicyTags(f.Tags)
+	if err != nil {
+		return errFn(fmt.Errorf("Error: '%v'", err))
```
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/sensors/tracing/generickprobe.go:695]: It seems other places are returning like that others are not, so not a big deal...

File: pkg/sensors/tracing/tags.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	t := fmt.Sprintf("%q", tag[:l])
+	newLen := len(t)
+	if newLen <= l || t[0] != '"' || t[newLen-1] != '"' {
+		return "", ErrTagSyntaxEscape
+	}
+
+	// Remove double quoted string so we pretty print it later in the events
+	return t[1 : newLen-1], nil
+}
```
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/sensors/tracing/tags.go:40]: I didn't bother check, just quickly ensuring that we escape and stay in boundary, will see, thanks

File: pkg/k8s/apis/cilium.io/v1alpha1/types.go [hunk 3]
```diff
@@ -248,6 +252,10 @@ type TracepointSpec struct {
 	// +kubebuilder:validation:Optional
 	// Selectors to apply before producing trace output. Selectors are ORed.
 	Selectors []KProbeSelector `json:"selectors,omitempty"`
+	// +kubebuilder:validation:Optional
+	// Tags to categorize the event, will be include in the event output.
+	// Maximum of 16 Tags are supported.
```
Comment (jrfastab) [inline pkg/k8s/apis/cilium.io/v1alpha1/types.go:257]: Nothing is enforcing this maxiumum at least as far as I can see here. So why bother with a max at all? It looks like a configured CRD value.

File: docs/content/en/docs/concepts/tracing-policy/tags.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+Events in this namespace relate to collect and export data about the internal system state.
+
+* "observability.filesystem": the event is about file system operations.
+* "observability.privilege": the event is about rights and permissions of a user or a process.
+* "observability.process": the event is about an instance of a Linux program being executed.
+
+## User defined Tags
+
+Users can define their own tags inside Tracing Policies. The official supported tags are documented
```
Comment (jrfastab) [inline docs/content/en/docs/concepts/tracing-policy/tags.md:69]: I'm not sure about this "official" supported tags. Tags can be anything I don't understand why we "support" a specific set of tags. I would just leave this as example tags.

File: pkg/sensors/tracing/tags.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	// Max tags of a Tracing Policy
+	TpMaxTags   = 16
+	TpMinTagLen = 2
+	TpMaxTagLen = 128
+)
+
+var (
+	ErrTagsSyntaxLong  = errors.New("tags field: too many tags")
+	ErrTagSyntaxShort  = errors.New("too short")
+	ErrTagSyntaxEscape = errors.New("escape failed")
```
[the remark this answers is not in the document]
Comment (jrfastab) [reply pkg/sensors/tracing/tags.go:21]: I don't see why we need tags defined by us here at all. If its just an array of strings we can just copy them through config through to the event. How the system database/SIEM/event monitor uses these doesn't need to be embedded inside Tetragon.

**Comment by jrfastab** [comment]
overall I agree tag fields are useful way for user to get a field they can use for DB or whatever event monitoring system for filtering, categorizing, etc. But, I'm against embedding specific tags in Tetragon. Why do we care?

**Comment by jrfastab** [comment]
Could you also add an export filter for tags? I think that would be useful to enable/disable certain types of events posting to JSON logs.

**Comment by tixxdz** [comment]
> overall I agree tag fields are useful way for user to get a field they can use for DB or whatever event monitoring system for filtering, categorizing, etc. But, I'm against embedding specific tags in Tetragon. Why do we care?

@jrfastab so we care:
1. As I'm pushing example https://github.com/cilium/tetragon/pull/1957 the privileges "raise" tracking, I want to tag all events with "observability.privilege" or "observability.privilege_escalation" if you prefer,  but we surely do want to groupe those.
2. We do want to have some standards and avoid developers pushing their own tags spread here and there...
3. With default shipped policy, users can quickly get those events knowing the standard tags, they won't guess! (but we still don't ship default yet, for now only examples and policylibrary)
4. For tracing policies that do enforcement we certainly won't tag those with "observability", so I want to avoid shipping say enforcement policies that may surprise users or block their workload by accident! we can in Tetragon daemon say load only "observability" policies for now, and users can augment on that, just avoid the selinux story of killing users's workload ;-)

So yeh I'm anticipating a bit, ok I will remove the default namespaces for now, let's get this merged first, then I think we will shortly have to update the policylibrary with default tags... in other words tell our users look for those tags and you get the corresponding events.  I let you sleep on it... :-) 

> Could you also add an export filter for tags? I think that would be useful to enable/disable certain types of events posting to JSON logs.

Ok will see how to add the export filter, fix and repush.

**Comment by jrfastab** [comment]
@tixxdz I'm all for groupings by tags and a default set of tags for different groups of observability and enforcement. I still don't follow why that needs to be embedded in the codebase. I'm happy with a dir in policylibrary for each taggroup though.

File: pkg/sensors/tracing/tags.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/sensors/tracing/tags.go:40]: hey @mtardy thanks I checked it and you are right fmt.Sprintf %q just ends up calling appendQuotedWith() which is basically the core of all these handling with the right parameters and we are sure to always have double quotes. Much appreciated ;-)  , changed.

File: pkg/sensors/tracing/tags.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	var newTags []string
+	for i, v := range tags {
+		_, ok := defaultTags[v]
+		if !ok {
+			parsed, err := escapeTag(v)
+			if err != nil {
+				return nil, fmt.Errorf("custom tag n%d: %v", i, err)
+			}
+			v = parsed
+		}
```
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/sensors/tracing/tags.go:69]: ok removed the default tags from code as requested by John

File: pkg/sensors/tracing/tags.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/sensors/tracing/tags.go:21]: @mtardy ok removed. Thanks!

@jrfastab ok removed for now the default tags from  code. Thanks!

**Comment by tixxdz** [comment]
> Thanks that's a good addition! I just have some remarks on the Go code that could be simplified I think to be easier to read.
> 
> I'm not fan of the naming `observability.privilege` though I don't think very talkative while privilege_escalation or whatever is from MITRE or other standards is more explicit.

Thanks @mtardy  for the review, and I changed it to `observability.privilege_escalation` ;-)

File: pkg/k8s/apis/cilium.io/v1alpha1/types.go (hunk 3 above)
Comment (tixxdz) [reply pkg/k8s/apis/cilium.io/v1alpha1/types.go:257]: So the enforcing is happening when parsing the crd, we don't have limits for other field and I'm not sure let me think

File: docs/content/en/docs/concepts/tracing-policy/tags.md (hunk 4 above)
Comment (tixxdz) [reply docs/content/en/docs/concepts/tracing-policy/tags.md:69]: So as we agreed some official tags are only in the doc to help keep things consistent, but I removed all the rest from the doc.

**Comment by tixxdz** [comment]
> @tixxdz I'm all for groupings by tags and a default set of tags for different groups of observability and enforcement. I still don't follow why that needs to be embedded in the codebase. I'm happy with a dir in policylibrary for each taggroup though.

So summary:
- Removed tags from the code, now they are only in documentation which I think having a default set of tags would help to keep things consistent.
- For export filter, I will open issue for it and we do it as a follow up.
- For enforcing Tags limit at kbuilder I'm not sure as this may change the type of field, hmm I am all for it so users don't push garbage into Tetragon , we can restrict the gRPC endpoint anyway but I'm not sure of the implication. hmm thinking again no argument against it I may change it and enforce it then at kbuilder

File: pkg/k8s/apis/cilium.io/v1alpha1/types.go (hunk 3 above)
Comment (jrfastab) [reply pkg/k8s/apis/cilium.io/v1alpha1/types.go:257]: My thought is if its just a slice does it matter what the max is?

**Review by jrfastab — APPROVED** [review]

**Comment by jrfastab** [comment]
Go ahead and merge after rebase to avoid future conflicts. Also I'll watch it today so when its green I can merge it as well if I see it.

**Pull request merged**